xarray/core/dataset.py

```diff
@@ -1955,7 +1955,35 @@ def fillna(self, value):
         -------
         Dataset
         """
-        out = self._fillna(value)
+        if utils.is_dict_like(value):
+            value_keys = value.data_vars.keys() if isinstance(value, Dataset)\
```

Comment:
maybe slightly cleaner to use duck typing, e.g., `value_keys = getattr(value, 'data_vars', value).keys()`